File: src/ast.rs
use serde::{Serialize};

#[derive(Debug, PartialEq, Serialize, Clone)]
pub enum Literal {
    IntLiteral(i64),
    FloatLiteral(f64),
    BoolLiteral(bool),
    CharLiteral(char),
    OctalLiteral(u32),
    HexLiteral(u8),
    BinaryLiteral(u32),
    StringLiteral(String),
}

#[derive(Debug, PartialEq, Serialize, Clone)]

pub enum Atom {
    Identifier(String),
    Literal(Literal),
    EndOfFile
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub enum Unary {
    SpreadExpr(Box<Expr>), // ...
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub enum BinaryOp {
    Assignment,
    Accessor,
    Invoke,  // can be called on any Body defines
    FieldDef,
    TypeDef,
    // BodyDef //  Body {} an invokable defined, has access to sibling values
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct Binary {
    pub op: BinaryOp,
    pub left: Box<Expr>,
    pub right: Box<Expr>,
}
#[derive(Debug, PartialEq, Serialize, Clone)]
pub enum TernaryOp {
    // Conditional,
    InvokeDefine   // func (func-params) {func-def}
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct Ternary {
    pub op: TernaryOp,
    pub left: Box<Expr>,
    pub middle: Box<Expr>,
    pub right: Box<Expr>,
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub enum GroupOp{
    AssignmentBlock,
    StatementBlock,
    ParamBlock,
    AnonymousType,
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub struct Group {
    pub op: GroupOp,
    pub exprs: Vec<Expr>,
}

#[derive(Debug, PartialEq, Serialize, Clone)]
pub enum Expr {
    // Atom
    Atom(Atom),             // Identifier, Literal, EndOfFile
    // Unary
    Unary(Unary),           // ...`a`
    // Binary
    Binary(Binary),         // `a` = `b`
    // Ternary
    Ternary(Ternary),       // `a` ? `b` : `c`  or  `a` ( `b` ) { `c` }
    // Grouping
    Group(Group),           // ( ... ) or { ... }
}

File: src/ast.nim
import 
    json,
    tables,
    strformat

type
  Expr = ref object
    case kind: ExprKind
    of Atom:
      atomVal: Atom
    of Unary:
      unaryVal: Unary
    of Binary:
      binaryVal: Binary
    of Group:
      groupVal: Group
    of TypeDef:
      typeDefVal: TypeDef
    of FuncCall:
      funcCallVal: FuncCall
    of EndOfFile: 
      discard

  LiteralKind = enum
    IntLiteral,
    FloatLiteral,
    BoolLiteral,
    CharLiteral,
    OctalLiteral,
    HexLiteral,
    BinaryLiteral,
    StringLiteral

  Literal = object
    case kind: LiteralKind
    of IntLiteral:
      intVal: int64
    of FloatLiteral:
      floatVal: float64
    of BoolLiteral:
      boolVal: bool
    of CharLiteral:
      charVal: char
    of OctalLiteral:
      octalVal: uint32
    of HexLiteral:
      hexVal: uint8
    of BinaryLiteral:
      binaryVal: uint32
    of StringLiteral:
      stringVal: string

  AtomKind = enum
    Identifier,
    Literal

  Atom = object
    case kind: AtomKind
    of Identifier:
      identifierVal: string
    of Literal:
      literalVal: Literal

  UnaryKind = enum
    SpreadExpr

  Unary = object
    kind: UnaryKind
    expr: Expr

  BinaryOp = enum
    Assignment,
    Accessor

  Binary = object
    op: BinaryOp
    left, right: Expr

  GroupOp = enum
    AssignmentBlock,
    StatementBlock,
    ParamBlock,
    AnonymousType

  Group = object
    op: GroupOp
    exprs: seq[Expr]

  TypeDef = object
    name: Expr
    fields: Expr

  FuncCall = object
    name: Expr
    args: Expr
    fields: Option[Expr]
    body: Option[Expr]

  ExprKind = enum
    Atom,
    Unary,
    Binary,
    Group,
    TypeDef,
    FuncCall,
    EndOfFile
